[Libxpm] ADD new port

vladimir-aubrecht opened the pull request:
**Describe the pull request**
This PR is adding Libxpm 3.5.11 port.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  windows, No

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  Yes, it follows the rules.

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

Neumann-A committed:
first batch of xwindow PR ports - up to dbus

Neumann-A committed:
first batch of xwindow PR ports - up to dbus

Neumann-A committed:
Merge branch 'xwindows_to_dbus' of https://github.com/Neumann-A/vcpkg into xwindows_to_dbus

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into xwindows_to_dbus

Neumann-A committed:
format manifest

Neumann-A committed:
version stuff

Neumann-A committed:
remove if block

Neumann-A committed:
version stuff

Neumann-A committed:
fix xcb hash

Neumann-A committed:
fix xproto dependency on meson blocking cross builds

Neumann-A committed:
adjust message

Neumann-A committed:
use X11_xcb_LIB

Neumann-A committed:
version stuff

Neumann-A committed:
put xlib arm64-windows on baseline

Neumann-A committed:
fix dbus on osx and linux

Neumann-A committed:
version stuff

Neumann-A committed:
forget to set the value

Neumann-A committed:
fix --export-dynamic on osx

Neumann-A committed:
version dbus

Neumann-A committed:
ci-retrigger

LilyWangLL committed:
Merge branch 'master' into xwindows_to_dbus

Commit:
Merge remote-tracking branch 'upstream/master' into xwindows_to_dbus

Commit:
add license (needs tool update)

Commit:
version adjustments

Commit:
version update

vladimir-aubrecht committed:
Added libxml port.

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into xwindows_to_dbus

Neumann-A committed:
Merge branch 'microsoft:master' into xwindows_to_dbus

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg into xwindows_to_dbus

Commit:
Merge branch 'master' of https://github.com/microsoft/vcpkg into xwindows_to_dbus

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into xwindows_to_dbus

Neumann-A committed:
remove unnecessary comments

Neumann-A committed:
v db

vladimir-aubrecht committed:
Merge branch 'microsoft:master' into master

vladimir-aubrecht committed:
Remove port version as it's initial port.

vladimir-aubrecht committed:
Added baseline version

vladimir-aubrecht committed:
Support only for windows and static

vladimir-aubrecht committed:
Allowed building debug version

vladimir-aubrecht committed:
update versions

Thomas1664 commented on ports/libxpm/CMakeLists.txt:
```diff
@@ -0,0 +1,70 @@
+)
+install(FILES ${LIBXPM_INC}
+        DESTINATION "${INSTALL_INC_DIR}"
+)
```
Please add a final newline

Thomas1664 commented on ports/libxpm/portfile.cmake:
```diff
@@ -0,0 +1,19 @@
+
+
+file(INSTALL "${SOURCE_PATH}/COPYRIGHT" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
```
here as well

vladimir-aubrecht replied:
Added new line.

Thomas1664 commented on ports/libxpm/portfile.cmake:
```diff
@@ -0,0 +1,19 @@
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS
+)
```
```suggestion
)
```

Thomas1664 commented on ports/libxpm/vcpkg.json:
```diff
@@ -0,0 +1,14 @@
+  "dependencies": [
+    {
+      "name": "vcpkg-cmake",
+      "host": true,
+      "platform": "windows"
```
```suggestion
      "host": true
```

Thomas1664 commented on ports/libxpm/vcpkg.json:
```diff
@@ -0,0 +1,14 @@
+  "version": "3.5.11",
+  "description": "XPM format pixmap library",
+  "homepage": "https://github.com/freedesktop/libXpm",
+  "supports": "windows & static",
```
```suggestion
  "supports": "windows",
```

Thomas1664 commented on ports/libxpm/portfile.cmake:
```diff
@@ -0,0 +1,19 @@
+vcpkg_from_github(
```
```suggestion
vcpkg_check_linkage(ONLY_STATIC_LIBRARY)

vcpkg_from_github(
```

You can prevent dll generation using `vcpkg_check_linkage`. This allows to build the port with triplets that build both static and dynamic libraries.

vladimir-aubrecht committed:
Update ports/libxpm/portfile.cmake

Co-authored-by: Thomas1664 <46387399+Thomas1664@users.noreply.github.com>

vladimir-aubrecht committed:
Added new lines

Neumann-A commented:
No reason to supply a CMakeLists.txt.....
Pc file is called just xpm not libxpm

vladimir-aubrecht committed:
Update ports/libxpm/vcpkg.json

Co-authored-by: Thomas1664 <46387399+Thomas1664@users.noreply.github.com>

vladimir-aubrecht committed:
Update ports/libxpm/portfile.cmake

Co-authored-by: Thomas1664 <46387399+Thomas1664@users.noreply.github.com>

vladimir-aubrecht committed:
Update ports/libxpm/vcpkg.json

Co-authored-by: Thomas1664 <46387399+Thomas1664@users.noreply.github.com>

vladimir-aubrecht committed:
Regenerated versions

vladimir-aubrecht commented:
Thank you @Thomas1664 for review! 😊

vladimir-aubrecht commented:
@Thomas1664 Do you know who could help me with merging it? 😊

Neumann-A commented on ports/libxpm/portfile.cmake:
```diff
@@ -0,0 +1,20 @@
+
+file(COPY "${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt" DESTINATION "${SOURCE_PATH}")
+
+vcpkg_cmake_configure(
```
There is no reason this uses CMake! With `xorg-macros` from #22642 this can simply call `vcpkg_configure_make`

Thomas1664 replied:
But make is significantly slower on Windows

Neumann-A replied:
but doesn't mean a maintenance burden for vcpkg. Install a pc file for use..... Also this doesn't do that many configure check so there is not that big of a time difference.

Neumann-A replied:
(also I have locally a working xpm port using make. Just saying.)

Thomas1664 replied:
My point is: Make is very slow on Windows because we have to spin up Mysys2. Furthermore, the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#prefer-using-cmake) encourages using CMake.

Neumann-A replied:
> When **multiple buildsystems** are available, prefer using CMake

When -> not the case here

> Additionally, **when appropriate**, it can be easier and more maintainable to rewrite alternative buildsystems into CMake

when appropriate -> not the case here. 
it can be easier and more maintainable -> not the case here


So if that `CMakeLists.txt` would be upstream that would be ok.

vladimir-aubrecht replied:
On my side I am fine with both options, up to you guys. From what I read in last comment getting rid off CMake seems to be the right path. @Thomas1664 are you ok with this?

(@Neumann-A Thank you for sample!)

JackBoosY replied:
Using custom CMakeLists is definitely risky:
1. Since it is unofficial, writing it according to other configuration files (such as configure/makefile) is likely to have bugs.
2. Unable to automatically sync with the version.
3. Features will hardly be complete.
4. Maintenance costs will be huge.

So I agree with @Neumann-A , we should use `vcpkg_configure_make` instead of `vcpkg_cmake_configure` and custom CMakeLists.txt here.

vladimir-aubrecht replied:
Thank you! I'll update the PR to switch to make 😊

Neumann-A commented on ports/libxpm/vcpkg.json:
```diff
@@ -0,0 +1,13 @@
+  "description": "XPM format pixmap library",
+  "homepage": "https://github.com/freedesktop/libXpm",
+  "supports": "windows",
+  "dependencies": [
```
according to https://gitlab.freedesktop.org/xorg/lib/libxpm/-/blob/master/xpm.pc.in#L9 this requires X11

vladimir-aubrecht replied:
I believe it's not required on Windows. Only place where I found reference to X11 for libxpm itself is in this [file](https://github.com/winlibs/libxpm/blob/76184718821c4788ab0cfc4ce6fc153cddebd917/src/XpmI.h) and there it has ifdef using simx.h instead of X11 header files in case of FOR_MSW symbol is defined.

Do you think it could cause some issues?

vladimir-aubrecht replied:
Resolving as dependency was needed and added after switch to make.

Neumann-A commented on ports/libxpm/CMakeLists.txt:
```diff
@@ -0,0 +1,70 @@
+project(libxpm)
+cmake_minimum_required(VERSION 2.8.12)
```
re-read https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html 
also version should be much higher

JonLiu1993 replied:
@vladimir-aubrecht,I see that @Neumann-A has some doubts here, I think you can explain it, thanks

Neumann-A commented on ports/libxpm/portfile.cmake:
```diff
@@ -0,0 +1,20 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO freedesktop/libXpm
+    REF 3425cbb0e6086f74783eafbe23df1121b655e006  #libXpm-v3.5.11
```
This is an outdated version.  Latest is 3.5.13

Neumann-A commented on ports/libxpm/portfile.cmake:
```diff
@@ -0,0 +1,20 @@
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+
+vcpkg_from_github(
```
vcpkg_from_gitlab(
    GITLAB_URL https://gitlab.freedesktop.org/xorg
    OUT_SOURCE_PATH SOURCE_PATH
    REPO lib/libxpm

???

JonLiu1993 commented on ports/libxpm/vcpkg.json:
```diff
@@ -0,0 +1,13 @@
+  "version": "3.5.11",
+  "description": "XPM format pixmap library",
+  "homepage": "https://github.com/freedesktop/libXpm",
+  "supports": "windows",
```
```suggestion
  "supports": "windows",
  "license": "MIT",
```

vladimir-aubrecht committed:
Update ports/libxpm/vcpkg.json

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

vladimir-aubrecht committed:
Libxpm is taken from gitlab now.

JonLiu1993 commented:
@vladimir-aubrecht, could you please reply to the review suggestion on @Neumann-A?

vladimir-aubrecht commented:
> @vladimir-aubrecht, could you please reply to the review suggestion on @Neumann-A?

I am sorry, could you point me which review suggestion do you have in mind? I thought I answered everything not related to CMake (that one I skipped as I am not sure if fixing CMake related comments make sense or not at this moment, considering @Neumann-A suggested to drop CMake and replace it with Make => I am waiting there on resolution between @Neumann-A  and @Thomas1664 to decide which way I should go).

JonLiu1993 commented on ports/libxpm/CMakeLists.txt:
```diff
@@ -0,0 +1,70 @@
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/include)
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/include/X11)
+
+if(BUILD_SHARED_LIBS)
+    add_library(libxpm SHARED ${LIBXPM_SRC})
+    add_definitions(-DLIBXPM_BUILD_DLL)
+else()
+    add_library(libxpm STATIC ${LIBXPM_SRC})
+endif()
```
@vladimir-aubrecht，and here

JonLiu1993 replied:
@vladimir-aubrecht ,I see that @Neumann-A  has some doubts here, I think you can explain it, thanks

vladimir-aubrecht replied:
Do I understand correctly that issue is about naming in the code? (I should replace all libxpm for "xpm"?)

JackBoosY replied:
> @Neumann-A Do I understand correctly that issue is about naming in the code? (I should replace all libxpm for "xpm"?)

This is meanless since you need to deprecate this file.

JonLiu1993 commented:
> > @vladimir-aubrecht, could you please reply to the review suggestion on @Neumann-A?
> 
> I am sorry @JonLiu1993, could you point me which review suggestion do you have in mind? I thought I answered everything not related to CMake (that one I skipped as I am not sure if fixing CMake related comments make sense or not at this moment, considering @Neumann-A suggested to drop CMake and replace it with Make => I am waiting there on resolution between @Neumann-A and @Thomas1664 to decide which way I should go).

@vladimir-aubrecht ,I see that @Neumann-A has some doubts here, I think you can explain it, thanks 
 https://github.com/microsoft/vcpkg/pull/25935#discussion_r927874384

vladimir-aubrecht commented:
> > > @vladimir-aubrecht, could you please reply to the review suggestion on @Neumann-A?
> > 
> > 
> > I am sorry @JonLiu1993, could you point me which review suggestion do you have in mind? I thought I answered everything not related to CMake (that one I skipped as I am not sure if fixing CMake related comments make sense or not at this moment, considering @Neumann-A suggested to drop CMake and replace it with Make => I am waiting there on resolution between @Neumann-A and @Thomas1664 to decide which way I should go).
> 
> @vladimir-aubrecht ,I see that @Neumann-A has some doubts here, I think you can explain it, thanks [#25935 (comment)](https://github.com/microsoft/vcpkg/pull/25935#discussion_r927874384)

Thank you for pointing it. I asked for some clarification and based on other replies I will either fix that code or drop it (based on result of CMake vs Make discussion).

JonLiu1993 commented:
@Neumann-A, can you explain the review comments?

Neumann-A commented:
@JonLiu1993: I think the main point here is that a decision of the team is required if either the x pr gets merged and this will wait for it or if you are ok with the cmakelist here which will be nuked if the x pr gets merged. Also the port name and the library name has to be checked against upstream naming.

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into xwindows_to_dbus

vladimir-aubrecht commented:
> @JonLiu1993: I think the main point here is that a decision of the team is required if either the x pr gets merged and this will wait for it or if you are ok with the cmakelist here which will be nuked if the x pr gets merged. Also the port name and the library name has to be checked against upstream naming.

@JonLiu1993 Could you please help with getting feedback from VCPKG team to get this decision? 😊

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

Neumann-A committed:
Apply suggestions from code review

Co-authored-by: Billy O'Neal <bion@microsoft.com>

Neumann-A committed:
Merge branch 'xwindows_to_dbus' of https://github.com/Neumann-A/vcpkg into xwindows_to_dbus

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into xwindows_to_dbus

Neumann-A committed:
nitpicks

Neumann-A committed:
rename xau to libxau

Neumann-A committed:
use vcpkg_install_copyright and silence usage

Neumann-A committed:
xtrans silence usage

Neumann-A committed:
format-manifest

Neumann-A committed:
v db

Neumann-A committed:
make license null for ports without exact match

Neumann-A committed:
xdmcp rename to libxdmcp

Neumann-A committed:
merge x11 wrapper into xlib
rename xlib to libx11 to avoid a metaport

Neumann-A committed:
Merge remote-tracking branch 'upstream/master' into xwindows_to_dbus

Neumann-A committed:
v db

Neumann-A committed:
missed dbus depending on x11.

Neumann-A committed:
v db

Neumann-A committed:
Update scripts/ci.baseline.txt

vladimir-aubrecht committed:
Initial switch to make

vladimir-aubrecht committed:
Merge branch 'main'

vladimir-aubrecht committed:
Merge branch 'main'

vladimir-aubrecht committed:
Merge branch 'xorg-macros'

JackBoosY commented:
Any progress here?

vladimir-aubrecht commented:
> Any progress here?

@JackBoosY sorry for delay, yes, I am still working on it just was busy also with other effort.

I made locally changes, but hit issue with enforced dependency on x11 in original project [configure.ac](https://github.com/freedesktop/libXpm/blob/master/configure.ac) file. @Neumann-A I saw you did some work around x11, do you know what is current status/plans for this to come to vcpkg?

Neumann-A commented:
> @Neumann-A I saw you did some work around x11, do you know what is current status/plans for this to come to vcpkg?

No. #22642 should be ready. Just waiting for the team to merge it.

Thomas1664 commented:
@vladimir-aubrecht @JonLiu1993 #22642 was merged.

JackBoosY commented:
I have no permission to push to this branch, can you please merge to master first?

Thanks.

vladimir-aubrecht committed:
Merge branch 'main'

vladimir-aubrecht committed:
Added required dependencies

vladimir-aubrecht commented:
> I have no permission to push to this branch, can you please merge to master first?
> 
> Thanks.

I merged with current master, added few missing dependencies, but it's still failing. I'll check it again tomorrow.

I also added you as contributor in case you'd like to do any changes.

vladimir-aubrecht committed:
Build fixed

vladimir-aubrecht committed:
Updated versions

vladimir-aubrecht committed:
Update version database

Commit:
format

Commit:
version

JackBoosY commented on ports/libxpm/portfile.cmake:
```diff
@@ -0,0 +1,39 @@
+     AUTOCONFIG
+ )
+
+#vcpkg_replace_string("${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/config.h"
+#        "#define STDC_HEADERS 1" "#define STDC_HEADERS 1\n\n#define FOR_MSW 1"
+# )
+
+#vcpkg_replace_string("${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/config.h"
+#        "#define STDC_HEADERS 1" "#define STDC_HEADERS 1\n\n#define FOR_MSW 1"
+# )
```
Is this not needed?

vladimir-aubrecht replied:
Yep, not needed. I dropped it.

vladimir-aubrecht committed:
Dropped not used code

vladimir-aubrecht committed:
versions

JackBoosY commented on ports/libxpm/vcpkg.json:
```diff
@@ -0,0 +1,17 @@
+  "description": "XPM format pixmap library",
+  "homepage": "https://github.com/freedesktop/libXpm",
+  "license": "MIT",
+  "supports": "windows",
```
One last question: I think this port is support Linux at begining, so what's the error on x64-Linux now?

vladimir-aubrecht replied:
I allowed just windows as I don't have environment prepared for linux cross-compiling and I am getting an error about missing cmake compiler. Basically my thinking was: currently we don't have libxpm at all, lets allow it for Windows and if somebody will need more environments, they can always create separate PR.

Neumann-A replied:
No, please enable linux. CI will simply test it. The question is if it should follow the same rules as the other x port having:
```
if(NOT X_VCPKG_FORCE_VCPKG_X_LIBRARIES AND NOT VCPKG_TARGET_IS_WINDOWS)
    message(STATUS "Utils and libraries provided by '${PORT}' should be provided by your system! Install the required packages or force vcpkg libraries by setting X_VCPKG_FORCE_VCPKG_X_LIBRARIES in your triplet")
    set(VCPKG_POLICY_EMPTY_PACKAGE enabled)
else()
 ...
endif()
```

vladimir-aubrecht replied:
Linux enabled and proposed check added.

vladimir-aubrecht replied:
@JackBoosY, @Neumann-A Do you miss anything else or can we merge it? 😊

Neumann-A commented on scripts/ci.baseline.txt:
```diff
@@ -1198,6 +1198,8 @@ xalan-c:x64-windows-static=fail
 xalan-c:arm64-windows=fail
 xerces-c:arm-uwp=fail
 xerces-c:x64-uwp=fail
+# Crosscompiling libx11 requires additional steps and patches to the Makefiles. 
+libx11:arm64-windows=fail
```
bad merge

vladimir-aubrecht replied:
Removed

vladimir-aubrecht committed:
Enabled support for linux

vladimir-aubrecht committed:
versions
